compute_tools/src/compute.rs

```diff
@@ -1937,19 +1937,46 @@ LIMIT 100",
         tokio::spawn(conn);
 
         // TODO: support other types of grants apart from schemas?
-        let query = format!(
-            "GRANT {} ON SCHEMA {} TO {}",
-            privileges
-                .iter()
-                // should not be quoted as it's part of the command.
-                // is already sanitized so it's ok
-                .map(|p| p.as_str())
-                .collect::<Vec<&'static str>>()
-                .join(", "),
-            // quote the schema and role name as identifiers to sanitize them.
-            schema_name.pg_quote(),
-            role_name.pg_quote(),
-        );
+
+        // check the role grants first - to gracefully handle read-replicas.
+        let select = "SELECT privilege_type
+            FROM pg_namespace
+                JOIN LATERAL (SELECT * FROM aclexplode(nspacl) AS x) acl ON true
+                JOIN pg_user users ON acl.grantee = users.usesysid
+            WHERE users.usename = $1
+                AND nspname = $2";
+        let rows = db_client
+            .query(select, &[role_name, schema_name])
+            .await
+            .with_context(|| format!("Failed to execute query: {select}"))?;
+
+        let mut already_granted = HashSet::new();
+        for row in rows {
+            let grant: String = row.get(0);
+            already_granted.insert(grant);
+        }
+
+        // check if all privileges are granted.
+        if privileges
+            .iter()
+            .all(|g| already_granted.contains(g.as_str()))
+        {
+            return Ok(());
+        }
+
+        let grants = privileges
+            .iter()
+            // should not be quoted as it's part of the command.
+            // is already sanitized so it's ok
+            .map(|p| p.as_str())
+            .collect::<Vec<&'static str>>()
```

Comment:
nit: May we use `itertools::join` here to avoid constructing a separate Vec?

Reply:
Also nit: if some of privileges may already be granted, maybe we should 
1. Take an iter of privileges
2. Filter out those contained in `already_granted`?
3. If iterator is non-empty, join using itertools and grant?